File: ruby/lib/onvo/dashboard/datasources.rb
# frozen_string_literal: true

require_relative './dashboard_resource'

# Dashboard Widget endpoints
class DashboardDatasources < DashboardResource
  # List out the datasources linked to a dashboard
  #
  # @param dashboard_id [String] The dashboard id to list datasources for
  # @return [Array] A list of all the datasources
  def list
    base_get("/dashboards/#{@dashboard_id}/datasources")
  end

  # Link a datasource to a dashboard
  #
  # @param datasource_id [String] The datasource id to link
  # @return [Hash] A dictionary containing the details of the new link between the dashboard and datasource
  def link(datasource_id)
    base_put("/dashboards/#{@dashboard_id}/datasources", body: { datasourceId: datasource_id })
  end

  # Unlink a datasource from a dashboard
  #
  # @param datasource_id [String] The datasource id to unlink
  # @return [Hash] A dictionary containing the details of the new link between the dashboard and datasource
  def unlink(datasource_id)
    base_delete("/dashboards/#{@dashboard_id}/datasources/#{datasource_id}")
  end
end

File: python/onvo/resources/dashboard_resources/datasources.py
from onvo.resources.dashboard_resources.base_resource import DashboardBaseResource


class DashboardDatasources(DashboardBaseResource):
    def list(self):
        """List all datasources linked to a dashboard

        Returns:
            list: A list of all the datasources
        """
        return self.base_get(f"/dashboards/{self.dashboard_id}/datasources")

    def link(self, datasource_id):
        """Link a datasource to a dashboard

        Args:
            datasource_id (str): The datasource id to link

        Returns:
            dict: A dictionary containing the details of the new link between the dashboard and datasource
        """
        return self.base_put(
            f"/dashboards/{self.dashboard_id}/datasources",
            data={"datasourceId": datasource_id},
        )

    def unlink(self, datasource_id):
        """Unlink a datasource from a dashboard

        Args:
            datasource_id (str): The datasource id to unlink

        Returns:
            dict: A dictionary that shows status of the unlink. Returns {"success": true} if successful.
        """
        return self.base_delete(
            f"/dashboards/{self.dashboard_id}/datasources/{datasource_id}"
        )
